autoload/vista/parser/lsp.vim

```diff
@@ -41,80 +43,103 @@ function! s:IsFileUri(uri) abort
-    let range = line.range
-    let lnum = range.start.line + 1
-    let col = range.start.character + 1
-    call add(a:container, {
-          \ 'lnum': lnum,
-          \ 'col': col,
-          \ 'kind': s:Kind2Symbol(line.kind),
-          \ 'text': line.name,
-          \ })
-    if has_key(line, 'children')
-      for child in line.children
-        call vista#parser#lsp#KindToSymbol(child, a:container)
-      endfor
+function! s:LspToLocalSymbol(sym, range)
+  return {
+    \ 'lnum': a:range.start.line + 1,
+    \ 'col': a:range.start.character + 1,
+    \ 'kind': s:Kind2Symbol(a:sym.kind),
+    \ 'text': a:sym.name,
+    \ }
+endfunction
+
+function! s:LocalToRawSymbol(sym)
+  return {
+    \ 'line': a:sym.lnum,
+    \ 'kind': a:sym.kind,
+    \ 'name': a:sym.text,
+    \ }
+endfunction
+
+function! s:IsDocumentSymbol(sym)
+  return has_key(a:sym, 'selectionRange')
+endfunction
+
+function! s:ParseSymbolInfoList(outlist, symbols) abort
+  for lspsym in a:symbols
+    let l:loc = lspsym.location
+    if s:IsFileUri(l:loc.uri)
+      call add(a:outlist, s:LspToLocalSymbol(lspsym, l:loc.range))
     endif
-  endif
+  endfor
 endfunction
 
-function! vista#parser#lsp#CocSymbols(symbol, container) abort
-  if vista#ShouldIgnore(a:symbol.kind)
-    return
-  endif
+function! s:ParseDocumentSymbolsRec(outlist, symbols, level) abort
+  for lspsym in a:symbols
+    let l:sym = s:LspToLocalSymbol(lspsym, lspsym.selectionRange)
+    let l:sym.level = a:level
+    call add(a:outlist, l:sym)
+    if has_key(lspsym, 'children')
+      call s:ParseDocumentSymbolsRec(a:outlist, lspsym.children, a:level + 1)
+    endif
+  endfor
+endfunction
 
-  let raw = { 'line': a:symbol.lnum, 'kind': a:symbol.kind, 'name': a:symbol.text }
-  call add(g:vista.raw, raw)
+function! s:GroupSymbolsByKind(symbols) abort
+  let l:groups = {}
+  for l:sym in a:symbols
+    let l:list = get(l:groups, l:sym.kind)
+    if !has_key(l:groups, l:sym.kind)
+      let l:list = []
+      let l:groups[l:sym.kind] = l:list
+    endif
+    call add(l:list, l:sym)
+  endfor
+  return l:groups
+endfunction
 
-  if a:symbol.kind ==? 'Method' || a:symbol.kind ==? 'Function'
-    call add(g:vista.functions, a:symbol)
+function! vista#parser#lsp#ParseDocumentSymbolPayload(resp) abort
+  let l:symbols = []
+  if s:IsDocumentSymbol(a:resp[0])
+    call s:ParseDocumentSymbolsRec(l:symbols, a:resp, 0)
+    call vista#parser#lsp#FilterDocumentSymbols(l:symbols)
+    call vista#parser#lsp#DispatchDocumentSymbols(l:symbols)
+    return l:symbols
+  else
+    call s:ParseSymbolInfoList(l:symbols, a:resp)
+    call vista#parser#lsp#FilterDocumentSymbols(l:symbols)
+    return s:GroupSymbolsByKind(l:symbols)
   endif
+endfunction
 
-  call add(a:container, {
-        \ 'lnum': a:symbol.lnum,
-        \ 'col': a:symbol.col,
-        \ 'text': a:symbol.text,
-        \ 'kind': a:symbol.kind,
-        \ 'level': a:symbol.level
-        \ })
+function! vista#parser#lsp#FilterDocumentSymbols(symbols) abort
```

Comment:
This function don't have to be public.

Reply:
DispatchDocumentSymbols and FilterDocumentSymbols were made public to allow to reuse them in actual CoC handler: the latter contains this exact code.

Reply:
hmm, where is the actual coc handler?

Reply:
`lsp.vim:133: vista#parser#lsp#CocSymbols`

Reply:
Well, not really: `CocSymbols` is to be replaced, handler is in `executive/coc.vim: s:Extract`